### PR TITLE
Support matching Args for SubgraphMatcher

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Subgraph matcher now handles the matching of non-Node arguments. 

Here are the 4 cases
- pn is Node, gn is Node: this go through the regular _match_node() function
- pn is Noed, gn is not a Node: this is a match if only pn is a placeholder op
- pn is not Node, gn is Node: this is a no match case
- pn is not a Node, gn is not a Node: this will go through the argument comparison. 

With this change
```
def target(x):
    return foo(x, 3)

def pattern(x, y):
    return foo(x, y) 
```

is a match

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85456

### CODE REVIEW

File: torch/fx/passes/utils/matcher_utils.py [hunk 1]
```diff
@@ -146,9 +167,23 @@ def _match_nodes(self, pn: Node, gn: Node, match: InternalMatch) -> bool:
...
-                           in zip(pn.all_input_nodes, gn.all_input_nodes)))
+        match_found = True
+
+        pn_flatten_args, _ = pytree.tree_flatten(pn.args)
+        gn_flatten_args, _ = pytree.tree_flatten(gn.args)
+
+        if len(pn_flatten_args) == len(gn_flatten_args):
+            for pn_, gn_ in zip(pn_flatten_args, gn_flatten_args):
+                if isinstance(gn_, Node) and isinstance(pn_, Node):
+                    matched = self._match_nodes(pn_, gn_, match)
```
Comment (jerryzh168) [inline torch/fx/passes/utils/matcher_utils.py:178]: nit: maybe we can just use `match_found` here

File: torch/fx/passes/utils/matcher_utils.py [hunk 2]
```diff
@@ -124,7 +125,27 @@ def _remove_overlapping_matches(self, matches: List[InternalMatch]) -> List[Inte
...
+                    return match.nodes_map[pn] == gn
+
+                match.nodes_map[pn] = gn
+                return True
+            else:
+                return False
+        elif not isinstance(pn, Node) and isinstance(gn, Node):
+            return False
+        else:
+            return type(gn) == type(pn) and gn == pn
```
Comment (jerryzh168) [inline torch/fx/passes/utils/matcher_utils.py:145]: does `gn == pn` imply `type(gn) == type(pn)`?

Comment (jerryzh168) [reply torch/fx/passes/utils/matcher_utils.py:145]: oh looks like not

**Review by jerryzh168 — APPROVED** [review]
thanks for the fix!

**Comment by jerryzh168** [comment]
also might be better to add a summary message

File: torch/fx/passes/utils/matcher_utils.py (hunk 1 above)
Comment (SherlockNoMad) [reply torch/fx/passes/utils/matcher_utils.py:178]: In this context, matched and match_found carries different meaning... 

`matched` is for each argument, `match_found` is for current node
if and only if `matched` for all the argument is True, then `match_found` is True.

**Comment by SherlockNoMad** [comment]
> also might be better to add a summary message

Thanks for reviewing this! 
I have updated summary message, merging now...

**Pull request closed without merging**